lib/resty/mlcache/ipc.lua

```diff
@@ -155,23 +171,68 @@ function _M:poll()
     for _ = self.idx, idx - 1 do
         self.idx = self.idx + 1
 
-        local v, err = self.dict:get(self.idx)
-        if err then
-            return nil, "failed to get event from shm: " .. err
-        end
+        -- fetch event from shm with a retry policy in case
+        -- we run our :get() in between another worker's
+        -- :incr() and :set()
+        -- 1: 0.001s
+        -- 2: 0.010s
+        -- 3: 0.100s
+        -- 4: 1.000s
+
+        local v
+
+        do
+            local perr
+            local pok         = true
+            local retry_delay = 0.0001
+
+            if not max_sleep then
+                max_sleep = 1
+            end
 
-        if not v then
-            return nil, "no event at index: " .. idx
+            while retry_delay < max_sleep do
+                v, err = self.dict:get(self.idx)
+                if err then
+                    log(ERR, "failed to get event from shm: ", err)
+                end
+
+                if v ~= nil or err then
+                    break
+                end
+
+                retry_delay = min(retry_delay * 10, max_sleep)
+
+                if pok then
+                    log(INFO, "no event data at index '", self.idx, "', ",
+                              "retrying in: ", retry_delay, "s")
+
+                    -- sleep is not available in all ngx_lua contexts
+                    -- if we fail once, never retry to sleep
+                    pok, perr = pcall(sleep, retry_delay)
```

Comment:
> The race condition you try to prevent here is from another worker. Within the same worker, the shm writes will always be atomic.

Certainly. But that doesn't prevent a user from potentially trying to poll from a non-yieldable context.